cg/meta/workflow/fastq.py

```diff
@@ -0,0 +1,275 @@
+                LOG.info("linking: %s -> %s", original_fastq_path, linked_fastq_path)
+                linked_fastq_path.symlink_to(original_fastq_path)
+            else:
+                LOG.debug("destination path already exists: %s", linked_fastq_path)
+
+        if not concatenate:
+            return
+
+        LOG.info("Concatenation in progress for sample %s.", sample_id)
+        for read, value in linked_reads_paths.items():
+            FastqHandler.concatenate(linked_reads_paths[read], concatenated_paths[read])
+            FastqHandler.remove_files(value)
+
+    @staticmethod
+    def parse_file_data(fastq_path: Path) -> dict:
+        with gzip.open(fastq_path) as handle:
+            header_line = handle.readline().decode()
+            header_info = FastqHandler.parse_header(header_line)
+
+            data = {
+                "path": fastq_path,
+                "lane": int(header_info["lane"]),
+                "flowcell": header_info["flowcell"],
+                "read": int(header_info["readnumber"]),
+                "undetermined": ("_Undetermined_" in fastq_path),
+            }
+            matches = re.findall(r"-l[1-9]t([1-9]{2})_", str(fastq_path))
+            if len(matches) > 0:
+                data["flowcell"] = f"{data['flowcell']}-{matches[0]}"
+            return data
+
+    @staticmethod
+    def create_fastq_name(
+        lane: str,
+        flowcell: str,
+        sample: str,
+        read: str,
+        date: dt.datetime = DEFAULT_DATE_STR,
+        index: str = DEFAULT_INDEX,
+        undetermined: Optional[str] = None,
+    ):
+        raise NotImplementedError
+
+
+class BalsamicFastqHandler(FastqHandler):
+    @staticmethod
+    def create_fastq_name(
+        lane: str,
+        flowcell: str,
+        sample: str,
+        read: str,
+        date: dt.datetime = DEFAULT_DATE_STR,
+        index: str = DEFAULT_INDEX,
+        undetermined: Optional[str] = None,
+    ) -> str:
+        """Name a FASTQ file following Balsamic conventions. Naming must be
+        xxx_R_1.fastq.gz and xxx_R_2.fastq.gz"""
+        flowcell = f"{flowcell}-undetermined" if undetermined else flowcell
+        date_str = date if isinstance(date, str) else date.strftime("%y%m%d")
+        return f"{lane}_{date_str}_{flowcell}_{sample}_{index}_R_{read}.fastq.gz"
+
+
+class MipFastqHandler(FastqHandler):
+    @staticmethod
+    def create_fastq_name(
+        lane: str,
+        flowcell: str,
+        sample: str,
+        read: str,
+        date: dt.datetime = DEFAULT_DATE_STR,
+        index: str = DEFAULT_INDEX,
+        undetermined: Optional[str] = None,
+    ) -> str:
+        """Name a FASTQ file following MIP conventions."""
+        flowcell = f"{flowcell}-undetermined" if undetermined else flowcell
+        date_str = date if isinstance(date, str) else date.strftime("%y%m%d")
+        return f"{lane}_{date_str}_{flowcell}_{sample}_{index}_{read}.fastq.gz"
+
+
+class MicrosaltFastqHandler(FastqHandler):
```

Comment:
Beautiful that you managed to merge the fastq handlers!!!

Reply:
feels good